pkg/catalog/catalog.go

```diff
@@ -65,66 +61,69 @@ func Exists(client *occlient.Client, componentType string) (bool, error) {
 func VersionExists(client *occlient.Client, componentType string, componentVersion string) (bool, error) {
 
 	// Retrieve the catalogList
-	catalogList, err := List(client)
+	// catalogs: map[imageName: string]map[namespace:string][]CatalogImage
+	catalogs, err := List(client)
 	if err != nil {
 		return false, errors.Wrapf(err, "unable to list catalog")
 	}
 
-	// Find the component and then return true if the version has been found
-	for _, supported := range catalogList {
-		if componentType == supported.Name {
-			// Now check to see if that version matches that components tag
-			for _, tag := range supported.Tags {
+	if catalogNSTags, ok := catalogs[componentType]; ok {
+		for _, tags := range catalogNSTags {
+			for _, tag := range tags {
 				if componentVersion == tag {
 					return true, nil
 				}
 			}
 		}
 	}
-
-	// Else return false if nothing is found
 	return false, nil
 }
 
 // getDefaultBuilderImages returns the default builder images available in the
-// openshift namespace
-func getDefaultBuilderImages(client *occlient.Client) ([]CatalogImage, error) {
-	imageStreams, err := client.GetImageStreams(occlient.OpenShiftNameSpace)
+// openshift namespace and current namespace
+// Return  map[imageName: string]map[namespace:string][]tags:string
+func getDefaultBuilderImages(client *occlient.Client) (map[string]map[string][]string, error) {
```

Comment:
uff this is pretty complicated return type. Wouldn't it be easier to create some nice struct for this?

Reply:
Iteration of struct array becomes tedious for any operations like search and delete.
But in a map the access is O(1) for most cases although at the cost of ease of code maintenance.
But it might now boil down to likely max size of list of imagestreams in current NS and openshift NS which looking at default size in my minishift deployment seems very small to bother about time complexities..

So,  I am changing https://github.com/redhat-developer/odo/pull/618/files#diff-4373e1ac6258d2d589cb87e96b1fc994R13 to 

```
type CatalogImage struct {
	Name string
	NamespacedTags map[Namespace: string]Tags: []string
}

```